Cannot Convert Partial<T> to Required<T>

Hi,

It's not currently possible to override types like so:

```javascript
import defaults from 'defaults';

type User = Partial<{
  name: string;
}>;

const defaultUser: Required<User> = {
  name: 'bob',
};

const user: User = {};

const requiredUser: Required<User> = defaults(user, defaultUser);
```

This will throw this error:

```bash
src/index.ts:13:7 - error TS2322: Type '{ name: string | undefined; }' is not assignable to type 'Required<Partial<{ name: string; }>>'.
  Types of property 'name' are incompatible.
    Type 'string | undefined' is not assignable to type 'string'.
      Type 'undefined' is not assignable to type 'string'.

13 const requiredUser: Required<User> = defaults(user, defaultUser);
```

Current workaround is to cast with `as Required<User>`, but I feel that the typing for this method should support this conversion.

--- a/index.d.ts
+++ b/index.d.ts
@@ -38,4 +38,8 @@ type DeepMerge<T, U> =
 		U extends Record<string, unknown> ?
 			{[K in keyof T | keyof U]: K extends keyof T ? K extends keyof U ? DeepMerge<T[K], U[K]> : T[K] : K extends keyof U ? U[K] : never}
 			: T
-		: U;
+		: U extends undefined ?
+			T extends undefined ?
+				U
+				: T
+			: U;

--- a/index.test-d.ts
+++ b/index.test-d.ts
@@ -45,3 +45,22 @@ import defaults from './index.js';
 	console.log(a.b); // Expect output to be 1
 	console.log(a.c); // Expect output to be 2
 }
+
+// Test that types get overriden by default when undefined
+{
+	type T = Partial<{
+		timeout: number;
+	}>;
+
+	const options: T = {};
+	const defaultOptions: Required<T> = {timeout: 100};
+	const result = defaults(options, defaultOptions);
+
+	// Expect result to be Required<T>
+	type Result = typeof result;
+	type Expected = Required<T>;
+	const _result: Expected = result;
+
+	// Safe to use now
+	console.log(result.timeout > 0);
+}

